### PR TITLE
Fix PSR4 in composer.json, add dump()

### PR DESCRIPTION
**Pull request by nicolas-grekas**
Empty psr-4 prefixes can force the autoloader to scan the `src/` directory for all namespaces. The best practice is to always make the prefix explicit.

`composer update` also, doing:
```
  - Updating twig/twig (v1.31.0 => v1.32.0) Loading from cache
  - Updating sensiolabs/security-checker (v4.0.0 => v4.0.1) Loading from cache
  - Updating phpunit/php-timer (1.0.8 => 1.0.9) Loading from cache
  - Updating phpunit/php-token-stream (1.4.9 => 1.4.11) Loading from cache
```

### CODE REVIEW

**Comment by stof** [comment]
👍

File: composer.json [hunk 1]
```diff
@@ -4,11 +4,15 @@
...
+        "psr-4": {
+            "AppBundle\\": "src/AppBundle/",
+            "CodeExplorerBundle\\": "src/CodeExplorerBundle/"
+        },
         "classmap": [ "app/AppKernel.php", "app/AppCache.php" ]
     },
     "autoload-dev": {
-        "psr-4": { "Tests\\": "tests/" }
+        "psr-4": { "Tests\\": "tests/" },
+        "files": [ "vendor/symfony/symfony/src/Symfony/Component/VarDumper/Resources/functions/dump.php" ]
```
Comment (theofidry) [inline composer.json]: why is this needed? isn't it supposed to already be included from the VarDumper `composer.json`?

**Comment by javiereguiluz** [comment]
Nice! I think this is important, so I've tweeted about it https://twitter.com/symfony_en/status/836232295329722370

File: composer.json [hunk 2]
```diff
@@ -4,11 +4,15 @@
     "type": "project",
     "description": "Symfony Demo Application",
     "autoload": {
-        "psr-4": { "": "src/" },
+        "psr-4": {
```
Comment (theofidry) [inline composer.json]: whilst this may boost performance, it may be harder for new users don't you think?

Comment (javiereguiluz) [inline composer.json]: But only if you create a ton of bundles in your app ... which by the way is now considered a bad practice in most of the cases!

Comment (theofidry) [inline composer.json]: Well to be honest I don't even see why you need any bundle here in the first place, but it's not limited to bundles. It highly depends on how you manage your `src` bundle. I would personally promote more something like having a `App` namespace for the whole `src`.

File: composer.json (hunk 1 above)
Comment (theofidry) [inline composer.json]: This is more as to ask for clarifications. I seem to be unable to use the `dump` function when the app is not bootstrapped sometimes so I took the habit to add a shortcut to make a FQCN call to avoid that problem, but this still looks weird as I would expect to be able to use `dump` at anytime.

Comment (nicolas-grekas) [inline composer.json]: When symfony/symfony is required, the dump function is not autoloaded. It is only when symfony/var-dumper is.

File: composer.json (hunk 2 above)
Comment (stof) [inline composer.json]: @theofidry the symfony-demo project has a AppBundle namespace for its source, because this is what the Symfony best practices are telling (and the demo is also about showing these best practices).

The extra CodeExplorerBundle is more a project-specific vendor bundle actually (we haven't shared it as a real vendor bundle, because this feature simply does not make sense outside the demo project).

and enw users won't find it harder, as they won't need to change the autoload config of symfony-demo.

**Comment by babaorum** [comment]
First of all, nice catch ! 
But after a few minutes on the subject, don't you think this is just a patch and not really correcting the real issue ? 
From my point of view we have composer on one hand which handle the autoloading. 
And we have symfony on the other hand which handle the application. 
But there is no way for symfony to tell composer where to look for our application directly. 
So before your PR, composer was kind of on is own to find our app. 
But now it's to the user to tell composer as well as symfony that we created a new bundle. 

I think the idealist way would be that composer has a way to load this from an external file, which would be generated by symfony on cache warm up for prod env. 

What do you think ? I would really appreciate your opinions on this. 

Thanks by advance. And once more, great job for digging this.

**Comment by nicolas-grekas** [comment]
@babaorum my humble opinion: we can wonder how the world could be better with more things being done out of the box, but that's really hard, many tried already with some success (composer is great!), and that's always improving. If you think you have some ideas that could make all of this even better, PRs are welcomed on both Composer and Symfony. For this very PR, it's unrelated. Just my 2cts :)

**Comment by juliendufresne** [comment]
The issue only exists  in the developper environment because when you push your code in production, you should use ˋcomposer install --optimize-autoloader` Or better `composer install --classmap-authoritativeˋ which will produce a classmap and is the fastest autoload.

File: composer.json (hunk 1 above)
Comment (Pierstoval) [inline composer.json]: Isn't the `dump` function loaded when `Debug::enable()` is called?

Comment (stof) [inline composer.json]: @Pierstoval no, it is loaded when the DebugBundle is booted. but when running tests, it is common for it not to be booted, making it painful. This is why we added it as a dev autoloading here for convenience.

**Comment by stof** [comment]
@juliendufresne ``--optimize-autoloader`` will not help for classmap misses. So while it is already better (PSR-4 fallbacks are checked after PSR-4 prefixes but before any PSR-0 config), it is still better to use a prefix.
This optimization is indeed useless when using ``--classmap-authoritative`` as composer will not do any PSR-4 lookup at runtime then.

File: composer.json (hunk 1 above)
Comment (Pierstoval) [inline composer.json]: Then symfony/symfony-standard#1025 should also be merged, shouldn't it? :)
